### PR TITLE
Histogram Rollout - Algorithms 30-35

### LINKED ISSUE #17669: Histogram rollout: Aglorithms 30-35
HistogramData rollout: algorithms <X to Y>

This issue is part of the HistogramData rollout, see https://github.com/mantidproject/mantid/issues/17641.
Algorithms covered in this issue are:

```Framework/Algorithms/src/RingProfile.cpp
Framework/Algorithms/src/RRFMuon.cpp
Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp
Framework/Algorithms/src/SANSDirectBeamScaling.cpp
Framework/Algorithms/src/SetUncertainties.cpp
Framework/Algorithms/src/SmoothData.cpp
```

### PR DESCRIPTION
**Pull request by DavidFair**
HistogramData rollout: algorithms 30-35
## Description of work.

This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:

```
Framework/Algorithms/src/RRFMuon.cpp
Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp
Framework/Algorithms/src/SANSDirectBeamScaling.cpp
Framework/Algorithms/src/SetUncertainties.cpp
Framework/Algorithms/src/SmoothData.cpp
```

Performance tests were added for:

```
SetUncertainties
SmoothData
```

The following algorithms have improved performance:

```
SetUncertainties (Approx 30%)
```
## To test.

Code review.

Fixes #17669.
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

### CODE REVIEW

File: Framework/Algorithms/inc/MantidAlgorithms/SampleCorrections/MayersSampleCorrectionStrategy.h [hunk 1]
```diff
@@ -61,15 +62,15 @@ class DLLExport MayersSampleCorrectionStrategy {
   /// Constructor
   MayersSampleCorrectionStrategy(
       MayersSampleCorrectionStrategy::Parameters params,
-      const std::vector<double> &tof, const std::vector<double> &sigIn,
-      const std::vector<double> &errIn);
+      const Mantid::HistogramData::Histogram &histogram);
```
Comment (LamarMoore) [inline Framework/Algorithms/inc/MantidAlgorithms/SampleCorrections/MayersSampleCorrectionStrategy.h]: Maybe use a more meaninful name. Like inputHist.

File: Framework/Algorithms/inc/MantidAlgorithms/SampleCorrections/MayersSampleCorrectionStrategy.h [hunk 2]
```diff
@@ -61,15 +62,15 @@ class DLLExport MayersSampleCorrectionStrategy {
...
   /// Destructor - defined in cpp file to use forward declaration with
   /// unique_ptr
   ~MayersSampleCorrectionStrategy();
 
   /// Return the correction factors
-  void apply(std::vector<double> &sigOut, std::vector<double> &errOut);
-  /// Calculate the self-attentation factor for a single mu*r value
+  void apply(Mantid::HistogramData::HistogramY &histoY,
```
Comment (LamarMoore) [inline Framework/Algorithms/inc/MantidAlgorithms/SampleCorrections/MayersSampleCorrectionStrategy.h]: Preserve the sigOut and errOut names.

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp [hunk 3]
```diff
@@ -134,9 +133,9 @@ void MayersSampleCorrection::exec() {
     params.sigmaSc = sampleMaterial.totalScatterXSection();
     params.cylRadius = radius;
     params.cylHeight = height;
-    MayersSampleCorrectionStrategy correction(params, inX, inputWS->readY(i),
-                                              inputWS->readE(i));
-    correction.apply(outputWS->dataY(i), outputWS->dataE(i));
+
+    MayersSampleCorrectionStrategy correction(params, inputWS->histogram(i));
+    correction.apply(outputWS->mutableY(i), outputWS->mutableE(i));
```
Comment (LamarMoore) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp]: Maybe apply could return a Histogram and you can use something like ouputWS->setHistogram(i, correction.apply());
You wouldn't then need the setSharedX above.

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 4]
```diff
@@ -53,11 +53,15 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
                    (double (*)(double))sqrt);
     std::generate(tof.begin(), tof.end(), Incrementer(100.0));
-    MayersSampleCorrectionStrategy mscat(createTestParameters(), tof, signal,
-                                         error);
 
+    Mantid::HistogramData::Histogram histo(tof, signal, error);
+
+    MayersSampleCorrectionStrategy mscat(createTestParameters(), histo);
+
+    Mantid::HistogramData::HistogramY signalHistogram(signal);
```
Comment (LamarMoore) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: you can use std::move here to move the data instead of copying. E.g. ```Mantid::HistogramData::HistogramY signalHistogram(std::move(signal));```

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 5]
```diff
@@ -22,9 +22,9 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
 
   void test_Attentuaton_Correction_For_Fixed_Mur() {
     std::vector<double> dummy(2, 0.0);
+    Mantid::HistogramData::Histogram histo(dummy, dummy);
```
Comment (LamarMoore) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: Why not something like ```Mantid::HistogramData::Histogram histo(Points(2, LinearGenerator(0, 1)), Counts(2, 0));``` or BinEdges if that is what's in use.

File: Framework/Algorithms/src/SetUncertainties.cpp [hunk 6]
```diff
@@ -116,25 +116,25 @@ void SetUncertainties::exec() {
...
+    outputWorkspace->setSharedY(i, inputWorkspace->sharedY(i));
     // copy the E or set to zero depending on the mode
     if (errorType.compare(ONE_IF_ZERO) == 0) {
-      outputWorkspace->dataE(i) = inputWorkspace->readE(i);
+      outputWorkspace->setSharedE(i, inputWorkspace->sharedE(i));
     } else {
-      outputWorkspace->dataE(i) =
-          std::vector<double>(inputWorkspace->readE(i).size(), 0.);
+      auto &errVals = outputWorkspace->mutableE(i);
+      std::fill(errVals.begin(), errVals.end(), 0);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/SetUncertainties.cpp]: You can simply write `outputWorkspace->mutableE(i) = 0.0;`.

File: Framework/Algorithms/src/SmoothData.cpp [hunk 7]
```diff
@@ -92,17 +92,17 @@ void SmoothData::exec() {
...
     // Now get references to the Y & E vectors in the input and output
     // workspaces
-    const MantidVec &Y = inputWorkspace->readY(i);
-    const MantidVec &E = inputWorkspace->readE(i);
-    MantidVec &newY = outputWorkspace->dataY(i);
-    MantidVec &newE = outputWorkspace->dataE(i);
+    const auto &Y = inputWorkspace->y(i);
+    const auto &E = inputWorkspace->e(i);
+    auto &newY = outputWorkspace->mutableY(i);
+    auto &newE = outputWorkspace->mutableE(i);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/SmoothData.cpp]: Move these lines below the `if (npts == 0) {` --- otherwise you are paying potential overhead even if you just hit the `continue` in the `if` and don't actually use the references.

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 8]
```diff
@@ -50,92 +48,120 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_DELTA(0.26514607, errVals.front(), delta);
+    TS_ASSERT_DELTA(0.2660792, errVals.back(), delta);
   }
 
   void
   test_Corrects_Both_Absorption_And_Multiple_Scattering_For_Histogram_Data() {
     using std::sqrt;
     const size_t nypts(100);
-    std::vector<double> signal(nypts, 2.0), tof(nypts + 1), error(nypts);
+    std::vector<double> signal(nypts, 2.0), tof(nypts), error(nypts);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: Why was this changed? According to the test name this should be testing for histogram data, i.e., bin edges and not points.

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 9]
```diff
@@ -50,92 +48,120 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
-    TS_ASSERT_DELTA(0.2660792, error.back(), delta);
+    TS_ASSERT_DELTA(0.26514310, errVals.front(), delta);
+    TS_ASSERT_DELTA(0.26607623, errVals.back(), delta);
   }
 
   void test_Corrects_For_Absorption_For_Histogram_Data() {
     using std::sqrt;
     const size_t nypts(100);
-    std::vector<double> signal(nypts, 2.0), tof(nypts + 1), error(nypts);
+    std::vector<double> signal(nypts, 2.0), tof(nypts), error(nypts);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: See above.

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 10]
```diff
@@ -50,92 +48,120 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_DELTA(1.6574675, errVals.front(), delta);
+    TS_ASSERT_DELTA(1.6609351, errVals.back(), delta);
   }
 
   // ---------------------- Failure tests -----------------------------
   void test_Tof_Not_Monotonically_Increasing_Throws_Invalid_Argument() {
     using std::sqrt;
     const size_t nypts(10);
-    std::vector<double> signal(nypts, 2.0), tof(nypts + 1), error(nypts);
+    std::vector<double> signal(nypts, 2.0), tof(nypts), error(nypts);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: See above.

**Review by SimonHeybrock — COMMENTED** [review]
Hi! Some small comments:

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp [hunk 11]
```diff
@@ -134,9 +131,10 @@ void MayersSampleCorrection::exec() {
     params.sigmaSc = sampleMaterial.totalScatterXSection();
     params.cylRadius = radius;
     params.cylHeight = height;
-    MayersSampleCorrectionStrategy correction(params, inX, inputWS->readY(i),
-                                              inputWS->readE(i));
-    correction.apply(outputWS->dataY(i), outputWS->dataE(i));
+
+    MayersSampleCorrectionStrategy correction(params, inputWS->histogram(i));
+    outputWS->setHistogram(i,
+                           correction.apply(std::move(inputWS->histogram(i))));
```
Comment (DavidFair) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp]: I looked at moving the entire class into a function but the class has several private methods which do the calculations for it. 
Looking at the usage stats the algorithm has been used around 10 times since 3.5 so I'm reluctant to completely tidy it however the changes (I'm yet to push) make it somewhat better.

Comment (SimonHeybrock) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp]: Ok, leave it is a class then but please do sufficient cleanup such that it is clear what it does. Right now it is not clear what the difference is between the histogram passed into the constructor and the histogram passed into `apply`. Can't we get rid of one of them?

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 12]
```diff
@@ -50,92 +48,120 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
+    const auto &tofVals = outHisto.x();
+    const auto &signalVals = outHisto.y();
+    const auto &errVals = outHisto.e();
 
     // Check some values
     const double delta(1e-06);
-    TS_ASSERT_DELTA(99.5, tof.front(), delta);
-    TS_ASSERT_DELTA(199.5, tof.back(), delta);
+    TS_ASSERT_DELTA(99.5, tofVals.front(), delta);
+    TS_ASSERT_DELTA(198.5, tofVals.back(), delta);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: @DavidFair Above you said if you switch to bin edges values are off by 1. Did you also change this back to the original state?

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp [hunk 13]
```diff
@@ -133,8 +133,7 @@ void MayersSampleCorrection::exec() {
     params.cylHeight = height;
 
     MayersSampleCorrectionStrategy correction(params, inputWS->histogram(i));
-    outputWS->setHistogram(i,
-                           correction.apply(std::move(inputWS->histogram(i))));
+    outputWS->setHistogram(i, correction.getCorrectedHisto());
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp]: :+1:

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrectionStrategy.cpp [hunk 14]
```diff
@@ -74,25 +74,23 @@ namespace Algorithms {
...
+      m_rng(new MersenneTwister(1)) {
+
+  double lastVal = m_histogram.x().front();
+  for (const auto &val : m_histogram.x())
+    if (val < lastVal) {
+      throw std::invalid_argument(
+          "TOF values are expected to be monotonically increasing");
+    } else {
+      lastVal = val;
+    }
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrectionStrategy.cpp]: I'd either leave this sanity check as it was, or fix it properly. As it is now you are still not testing what the error message claims (i.e., monotonically increasing values).
The plan is to have such a validation as part of the `BinEdges` class one day, so I'd go for the easy option and just leave things as they were.

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrectionStrategy.cpp [hunk 15]
```diff
@@ -342,7 +338,7 @@ double MayersSampleCorrectionStrategy::sigmaTotal(const double flightPath,
  * @return The associated TOF value
  */
 double MayersSampleCorrectionStrategy::tof(const size_t i) const {
-  return m_histogram ? 0.5 * (m_tof[i] + m_tof[i + 1]) : m_tof[i];
+  return m_histogram.points()[i];
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrectionStrategy.cpp]: This makes things slower: If `m_histogram` is storing `BinEdges`, `points()` will now be recomputed once for every `i` (the whole vector!). You should buffer the points in the class to avoid this.

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 16]
```diff
@@ -80,22 +87,29 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
     const double delta(1e-06);
-    TS_ASSERT_DELTA(99.5, tof.front(), delta);
-    TS_ASSERT_DELTA(199.5, tof.back(), delta);
+    TS_ASSERT_DELTA(99.5, tofVals.front(), delta);
+    TS_ASSERT_DELTA(199.5, tofVals.back(), delta);
 
-    TS_ASSERT_DELTA(0.37497317, signal.front(), delta);
-    TS_ASSERT_DELTA(0.37629282, signal.back(), delta);
+    TS_ASSERT_DELTA(0.37497317, signalVals.front(), delta);
+    TS_ASSERT_DELTA(0.37629281, signalVals.back(), delta);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: Value changed?

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 17]
```diff
@@ -80,22 +87,29 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
 
-    TS_ASSERT_DELTA(0.26514607, error.front(), delta);
-    TS_ASSERT_DELTA(0.2660792, error.back(), delta);
+    TS_ASSERT_DELTA(0.26514607, errVals.front(), delta);
+    TS_ASSERT_DELTA(0.26607920, errVals.back(), delta);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: Value changed?

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 18]
```diff
@@ -107,35 +121,47 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
     // Check some values
     const double delta(1e-06);
-    TS_ASSERT_DELTA(99.5, tof.front(), delta);
-    TS_ASSERT_DELTA(199.5, tof.back(), delta);
+    TS_ASSERT_DELTA(99.5, tofVals.front(), delta);
+    TS_ASSERT_DELTA(199.5, tofVals.back(), delta);
 
-    TS_ASSERT_DELTA(2.3440379, signal.front(), delta);
-    TS_ASSERT_DELTA(2.3489418, signal.back(), delta);
+    TS_ASSERT_DELTA(2.3440378, signalVals.front(), delta);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: Value changed?

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 19]
```diff
@@ -107,35 +121,47 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_DELTA(2.3489418, signalVals.back(), delta);
 
-    TS_ASSERT_DELTA(1.6574851, error.front(), delta);
-    TS_ASSERT_DELTA(1.6609527, error.back(), delta);
+    TS_ASSERT_DELTA(1.6574850, errVals.front(), delta);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: Value changed?

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h [hunk 20]
```diff
@@ -107,35 +121,47 @@ class MayersSampleCorrectionStrategyTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_DELTA(1.6609527, errVals.back(), delta);
   }
 
   // ---------------------- Failure tests -----------------------------
   void test_Tof_Not_Monotonically_Increasing_Throws_Invalid_Argument() {
     using std::sqrt;
     const size_t nypts(10);
-    std::vector<double> signal(nypts, 2.0), tof(nypts + 1), error(nypts);
+    std::vector<double> signal(nypts, 2.0), tof(nypts), error(nypts);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: Undo this as well?

**Review by SimonHeybrock — COMMENTED** [review]
@DavidFair Good, looks much better now. Just a few small things left to fix:

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h (hunk 12 above)
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: Sorry, I didn't quite get that. Are you saying that the unit test fails if you put the unit test into its original state (apart from packing X, Y, and E into a histogram)?

File: Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h (hunk 16 above)
Comment (SimonHeybrock) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: It was also bin edges before you changed anything, as far as I can see? So why would the values change?

Comment (DavidFair) [inline Framework/Algorithms/test/MayersSampleCorrectionStrategyTest.h]: I'm not sure. However previously `MayersSampleCorrectionStrategyTest` took in raw vectors and manipulated them directly so I assumed by changing the type to Histogram it resolved some underlying bug in how it handled point/bin data. 
However I cannot find any changes which explain what caused these values to change in any of my changes as they should all be superficial except for the fact we have the interface the enforce bin edges or points.

**Comment by SimonHeybrock** [comment]
@DavidFair I reverted the changes to the test result values and tests pass on my system, so I don't know why they don't on yours. Let us see what the build servers say.

I also did some other small cleanup in the test. Sorry for hijacking your branch.

**Comment by DavidFair** [comment]
No thats perfectly fine - I've been completely stumped the last few days on it so I'm glad to see it working and finally know it was probably an issue at my end vs what changes had been completed.

**Comment by SimonHeybrock** [comment]
@LamarMoore Since I made changes to this PR I shouldn't be the one putting the shipit, could you have a look (at least at the parts I changed) and see if it is ready?

**Comment by martyngigg** [comment]
@LamarMoore Is this ready to go?

**Comment by LamarMoore** [comment]
:shipit:
